torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,724 @@
+        Py_DECREF(i.self_);
+    }
+    module_calls_.clear();
+}
+
+void PythonTracer::record_py_call(TraceContext* ctx, PyFrameObject* frame) {
+    events_.emplace_back(TraceTag::kPy_Call, frame->f_lasti, ctx, frame->f_code);
+    store_description(frame);
+    track_module(frame);
+}
+
+void PythonTracer::record_c_call(TraceContext* ctx, PyFrameObject* frame, PyObject* arg) {
+    events_.emplace_back(TraceTag::kC_Call, frame->f_lasti, ctx, arg);
+}
+
+void PythonTracer::record_return(TraceContext* ctx, PyFrameObject* frame, TraceTag tag) {
+    events_.emplace_back(tag, frame->f_lasti, ctx);
+}
+
+// NB:
+//  `frame->f_lasti` will advance as the interpreter progresses through the
+//  code object. Thus, we need to call `store_description` when we record the
+//  call rather than the return. (Otherwise we would get the line with the
+//  return stmt.)
+void PythonTracer::store_description(PyFrameObject* frame) {
+    const auto& it = code_descriptions_.find({ frame->f_code, frame->f_lasti });
+    if C10_UNLIKELY(it == code_descriptions_.end()) {
+        code_descriptions_.insert({
+            { frame->f_code, frame->f_lasti },
+            {
+                /*line_no=*/ PyCode_Addr2Line(frame->f_code, frame->f_lasti),
+                /*filename=*/ THPUtils_unpackString(frame->f_code->co_filename),
+                /*funcname=*/ THPUtils_unpackString(frame->f_code->co_name)
+            }
+        });
+    }
+}
+
+void PythonTracer::track_module(PyFrameObject* frame) {
+    if ((PyObject*)(frame->f_code) == module_call_code_) {
+        // By default, CPython stores locals in a "fast" format, with an array
+        // of names and an array of values. Consequently, frame->f_locals is
+        // NULL since the interpreter has no need to populate it.
+        //
+        // If these arrays were part of the public API then we could very
+        // quickly access `self`. Unfortunately they are not, and moreover are
+        // not stable across versions. As a result, we are forced to call
+        // `PyFrame_FastToLocals` which forces the interpreter to materialize
+        // the full dict of locals.
+        PyFrame_FastToLocals(frame);
+        auto self = PyDict_GetItemString(frame->f_locals, "self");
+        Py_INCREF(self);
+        module_calls_.emplace_back(
+            /*event_index=*/events_.size() - 1,
+            /*self=*/self
+        );
+        PyFrame_LocalsToFast(frame, 0);
+    }
+};
+
+
+// ============================================================================
+// == Post processing =========================================================
+// ============================================================================
+
+class PyTraceReplay {
+ public:
+    static std::vector<std::unique_ptr<PyTraceEvent>> get_events() {
+        return PyTraceReplay().replay_stack();
+    }
+
+ private:
+    PyTraceReplay();
+    std::vector<std::unique_ptr<PyTraceEvent>> replay_stack() const;
+
+    struct ReplayFrame {
```

Comment:
Can you not populate PyTraceEvent directly and keep a little extra on the side?

Reply:
Yeah, I should. I'll change it.

Reply:
Feel free to do it later in a follow-up diff.

Reply:
Or perhaps we can return to this once the kineto refactoring diffs have landed and perhaps we can populate the fina trace activities directly.

Reply:
Pulling the duplicate fields out of `ReplayFrame` and just adding a shared_ptr field turned out to be pretty trivial. Longer term I agree that a lot of this is fairly generic and closer integration with kineto is in order.